src/libraries/ReservoirLibrary.sol

```diff
@@ -0,0 +1,224 @@
+pragma solidity 0.8.13;
+
+import { ERC20 } from "solmate/tokens/ERC20.sol";
+
+import { IReservoirPair } from "v3-core/src/interfaces/IReservoirPair.sol";
+import { IGenericFactory } from "v3-core/src/interfaces/IGenericFactory.sol";
+import { ExtraData } from "src/interfaces/IReservoirRouter.sol";
+
+import { ConstantProductPair } from "v3-core/src/curve/constant-product/ConstantProductPair.sol";
+import { StablePair } from "v3-core/src/curve/stable/StablePair.sol";
+
+import { StableMath } from "v3-core/src/libraries/StableMath.sol";
+
+library ReservoirLibrary {
+    uint256 public constant FEE_ACCURACY  = 10_000;
+
+    // returns sorted token addresses, used to handle return values from pairs sorted in this order
+    function sortTokens(address tokenA, address tokenB) internal pure returns (address token0, address token1) {
+        require(tokenA != tokenB, "RL: IDENTICAL_ADDRESSES");
+        (token0, token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
+        require(token0 != address(0), "RL: ZERO_ADDRESS");
+    }
+
+    // calculates the CREATE2 address for a pair without making any external calls
+    function pairFor(address factory, address tokenA, address tokenB, uint256 curveId) internal pure returns (address pair) {
+        (address token0, address token1) = sortTokens(tokenA, tokenB);
+
+        bytes memory lInitCode;
+
+        if (curveId == 0) {
+            lInitCode = abi.encodePacked(type(ConstantProductPair).creationCode, abi.encode(token0, token1));
+        }
+        else if (curveId == 1) {
+            lInitCode = abi.encodePacked(type(StablePair).creationCode, abi.encode(token0, token1));
+        }
+        else {
+            revert("RL: CURVE_DOES_NOT_EXIST");
+        }
+
+        pair = address(uint160(uint256(keccak256(abi.encodePacked(
+                bytes1(0xff),
+                factory,
+                bytes32(0),
+                keccak256(lInitCode)
```

Comment:
What's the gas cost for doing this, surely it's massive? If it is, we should document it with natspec & make sure we don't call this function anywhere onchain (just in view methods)

Reply:
I did a test, purely calling that function. 

Gas cost was 
`68172`

In contrast if we call `factory.getPair`, the gas cost it
`12206`

Wow this is really surprising to me. 

I guess when UniV2 implemented it with the salt, it made sense (cheaper to calculate than make external call), but the way we calculate the init code makes it more expensive?

Reply:
Our initCodeHash is not stable because of immutables hence we must hash the code every time

Reply:
will address in #2, resolving for now